### PR TITLE
Use storaged 2.0

### PR DESCRIPTION
**Pull request by mvollmer**
To test this, you need to install storaged 2.1.1 or later.

```
dnf copr enable phatina/storaged
dnf install storaged
dnf install storaged-lvm2
```

Then you need to log out and in again.

### CODE REVIEW

**Comment by mvollmer** [comment]
Now it uses base1/patterns.

**Comment by mvollmer** [comment]
Let's wait for a new storaged release that fixes https://bugs.freedesktop.org/show_bug.cgi?id=91011

**Comment by mvollmer** [comment]
Rebased and bumped tag so that we pick up storaged 2.1.1.

**Comment by mvollmer** [comment]
Let's remove the HACK re https://github.com/storaged-project/storaged/issues/12

File: pkg/storaged/storage.css [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    padding-right: 5px;
+}
+
+#storage .storage-mounts .table {
+    table-layout: fixed;
+}
+
+#storage .storage-mounts .table th,
+#storage .storage-mounts .table td {
+  white-space: nowrap;
```
Comment (stefwalter) [inline pkg/storaged/storage.css]: Identation.

File: test/guest/cockpit-fedora-22.setup [hunk 2]
```diff
@@ -9,7 +9,7 @@ echo foobar | passwd --stdin root
 # the distribution repository is updated.
 #
 
-COCKPIT_DEPS="udisks2 json-glib realmd lvm2 mdadm glib-networking libssh selinux-policy-targeted docker pcp-libs pcp kubernetes etcd subscription-manager"
+COCKPIT_DEPS="udisks2 json-glib realmd lvm2 mdadm glib-networking libssh selinux-policy-targeted docker pcp-libs pcp kubernetes etcd subscription-manager storaged storaged-lvm2"
```
Comment (stefwalter) [inline test/guest/cockpit-fedora-22.setup]: Can we remove the direct lvm2 dep? Both here and in cockpit.spec?

File: tools/cockpit.spec [hunk 3]
```diff
@@ -376,6 +364,17 @@ cluster. Installed on the Kubernetes master. This package is not yet complete.
 
 %endif
 
+%package storaged
+Summary: Cockpit user interface for storage, using Storaged
+Requires: mdadm
+Requires: lvm2
```
Comment (stefwalter) [inline tools/cockpit.spec]: Do we still have these direct deps?

File: test/storagelib.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+    # Deep inspection
+
+    # The devices.js scope allows us to execute code inside of it.  We
+    # use that to wait for the internal state to reach an expected
+    # point before opening dialogs that depend on that state.
+
+    def wait_devices_func(self, code, *args):
+        self.browser.wait_js_func("""function (code, args) {
+            return window.phantom_eval_in_devices("(" + code + ")").apply(this, args);
```
Comment (stefwalter) [inline test/storagelib.py]: This seems quite hacky. Shouldn't we just tag the displayed tags with semantic attributes, and not need to have this extremely fragile code poking around inside the internals of the storage page?

**Comment by stefwalter** [comment]
devices.js is too long for github to allow review on it.

**Comment by stefwalter** [comment]
The following functions need formatting fixing: fmt_size, fmt_size_long, fmt_rate

**Comment by stefwalter** [comment]
Why does fmt_part use mustache.render instead of the simpler cockpit.format()?

**Comment by stefwalter** [comment]
>  cockpit.spawn(["date", "+%s"])

Do we need to do this globally somehow? Will each page end up doing this?

**Comment by stefwalter** [comment]
>  function wait_all(objects, callback) {

This function has risky semantics ... eg: modifying its arguments ... suggest moving it to the site where it's used in that private function scope.

**Comment by stefwalter** [comment]
>     /* As a special service, we try to also show UDisks2 jobs.  As
>     * a shortcut, we assume that we can blindly transform a
>     * UDisks2 object path into a Storaged object path for the
>     * same object.
>     */

Does this pull request support udisks2? If not, why are we doing this?

**Comment by stefwalter** [comment]
Why are the dialogs written in JSON/javascript? Is there a good reason? In general we're trying to do this with HTML/CSS and moving away from the jQuery spaghetti we had before.

**Comment by mvollmer** [comment]
> Let's remove the HACK re storaged-project/storaged#12

Done.

File: test/storagelib.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        # VirtIO devices don't seem to appear early enough for udev to
+        # pick them up reliably.  We just wait a bit.
+        #
+        self.machine.write("/usr/lib/udev/rules.d/59-fixup-serial.rules",
+             'SUBSYSTEM=="block" KERNEL=="vd*" IMPORT{program}="/bin/sh -c \'sleep 0.5; s=$(cat /sys/block/$(basename $tempnode)/serial); echo ID_SERIAL=$s\'"')
+        self.machine.execute("udevadm trigger");
+
+        # HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1225905
+        # HACK - https://bugzilla.redhat.com/show_bug.cgi?id=1225957
+        #
```
Comment (mvollmer) [inline test/storagelib.py]: These two hacks are only necessary on Rawhide.

File: test/storagelib.py (hunk 4 above)
Comment (mvollmer) [inline test/storagelib.py]: I agree.

**Comment by mvollmer** [comment]
> Why does fmt_part use mustache.render instead of the simpler cockpit.format()?

Good question.  I think I wanted to use Mustache to esacpe things for us properly, but fmt_part doesn't need any escaping.

**Comment by mvollmer** [comment]
> The following functions need formatting fixing: fmt_size, fmt_size_long, fmt_rate

Fixed.

File: test/storagelib.py (hunk 4 above)
Comment (mvollmer) [inline test/storagelib.py]: Fixed, hopefully.

File: pkg/storaged/devices.html [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+
+  <script id="mounts-tmpl" type="x-template/mustache">
+    <div class="panel panel-default storage-mounts">
+      <div class="panel-heading">
+        <span translatable="yes">Filesystems</span>
+      </div>
+      <table class="table table-hover">
+        <thead>
+          <tr>
+            <th class="mount-name">Name</th>
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

File: pkg/storaged/devices.html [hunk 7]
```diff
@@ -0,0 +1,675 @@
...
+            {{#Cancelable}}<button translatable="yes" class="btn btn-default" data-action="job_cancel" data-arg="{{path}}">Cancel</button>{{/Cancelable}}
+          </td>
+        </tr>
+        {{/Jobs}}
+      </table>
+    </div>
+    {{/HasJobs}}
+  </script>
+
+  <div id="unsupported" hidden>
```
Comment (petervo) [inline pkg/storaged/devices.html]: Should we use curtains/blank-slate for this rather than alerts?

File: pkg/storaged/devices.html [hunk 8]
```diff
@@ -0,0 +1,675 @@
...
+  <div id="storage" hidden>
+    <div class="col-md-8 col-lg-9">
+      <div id="storage-graph-toolbar" class="zoom-controls" style="text-align:right;margin-bottom:10px">
+        <div class="dropdown" style="display:inline-block">
+          <button class="btn btn-default dropdown-toggle" data-toggle="dropdown">
+            <span style="width:6em;text-align:left;padding-left:5px;display:inline-block"></span>
+            <span class="caret"></span>
+          </button>
+          <ul class="dropdown-menu" role="menu">
+            <li role="presentation"><a role="menuitem" tabindex="-1" data-action="goto-now">Go to now</a></li>
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

File: pkg/storaged/devices.html [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+          <span class="glyphicon glyphicon-zoom-out"></span>
+        </button>
+        <div class="btn-group">
+          <button class="btn btn-default fa fa-angle-left" data-action="scroll-left"></button>
+          <button class="btn btn-default fa fa-angle-right" data-action="scroll-right"></button>
+        </div>
+      </div>
+      <div class="row">
+        <div class="col-sm-6">
+          <div>Reading<span style="float:right" id="storage-reading-text"></span></div>
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

File: pkg/storaged/devices.html [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+          <div style="height:120px" id="storage-reading-graph" class="zoomable-plot"></div>
+        </div>
+        <div class="col-sm-6">
+          <div>Writing<span style="float:right" id="storage-writing-text"></span></div>
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

File: pkg/storaged/devices.html [hunk 11]
```diff
@@ -0,0 +1,675 @@
...
+          <td translatable="yes" context="storage">Capacity</td>
+          <td>{{Size}}</td>
+        </tr>
+      </table>
+      {{/VGroup}}
+    </div>
+    <div class="panel panel-default" id="pvols">
+      <div class="panel-heading">
+        <span translatable="yes">Physical Volumes</span>
+        <button style="float:right" class="btn btn-default storage-privileged"
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

File: pkg/storaged/devices.html [hunk 12]
```diff
@@ -0,0 +1,675 @@
...
+        </li>
+        {{/PVols}}
+      </ul>
+    </div>
+    {{{Content}}}
+  </script>
+
+  <div id="storage-detail" class="container-fluid" hidden>
+    <ol class="breadcrumb">
+      <li><a>Storage</a></li>
```
Comment (petervo) [inline pkg/storaged/devices.html]: translatable?

**Comment by petervo** [comment]
devices.js is a little hard to review as it's too big for github. I don't know if it's possible (or worth it) to split it up. I noticed a couple indentation issues but other than that it seemed ok. I was able to run through some basic functionality without problems.

@dperpeet do you want to take a look at this as well.

**Comment by mvollmer** [comment]
> devices.js is a little hard to review as it's too big for github. I don't know if it's possible (or worth it) to split it up.

I could try to split out the D-Bus stuff from the UI stuff, maybe by making a StoragedClient.  I think that would be worth it.

**Comment by dperpeet** [comment]
> I could try to split out the D-Bus stuff from the UI stuff, maybe by making a StoragedClient. I think that would be worth it.

I think you could go even further. Your "utilities" could be separate, D-Bus, and maybe even split up the pages ("Overview"). Bundled it doesn't impact performance really, but makes it a lot more readable.

**Comment by mvollmer** [comment]
Let's wait for storaged 2.1.2, to hopefully fix the check-storage-mdraid failure.

**Comment by mvollmer** [comment]
> Why does fmt_part use mustache.render instead of the simpler cockpit.format()?

Fixed.

File: pkg/storaged/storage.css (hunk 1 above)
Comment (mvollmer) [inline pkg/storaged/storage.css]: Fixed

File: test/guest/cockpit-fedora-22.setup (hunk 2 above)
Comment (mvollmer) [inline test/guest/cockpit-fedora-22.setup]: Yes, and also mdadm.  Done.

File: tools/cockpit.spec (hunk 3 above)
Comment (mvollmer) [inline tools/cockpit.spec]: Done.

**Comment by mvollmer** [comment]
> Do we need to do this globally somehow? Will each page end up doing this?

That's a good idea.

**Comment by mvollmer** [comment]
> >  function wait_all(objects, callback) {

> This function has risky semantics ... eg: modifying its arguments ... suggest moving it to the site where it's used in that private function scope.

Good point.  Done.

**Comment by mvollmer** [comment]
> Does this pull request support udisks2? If not, why are we doing this?

It supports having udisks2 installed.  If someone runs a job via UDisks2, we show it and allow it to be cancelled.

But, yeah, that's maybe not important.

**Comment by mvollmer** [comment]
> Why are the dialogs written in JSON/javascript? Is there a good reason?

I seem to like this approach. :-)

Can you give a quick example of how they would look in HTML/CSS?

File: test/storagelib.py (hunk 5 above)
Comment (mvollmer) [inline test/storagelib.py]: Fixed.

File: pkg/storaged/devices.html (hunk 6 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed.

File: pkg/storaged/devices.html (hunk 8 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed.

File: pkg/storaged/devices.html (hunk 9 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed.

File: pkg/storaged/devices.html (hunk 10 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed.

File: pkg/storaged/devices.html (hunk 11 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed

File: pkg/storaged/devices.html (hunk 12 above)
Comment (mvollmer) [inline pkg/storaged/devices.html]: Fixed.

**Comment by mvollmer** [comment]
Now the structure should be clearer, although the modules are probably not fit to be reused in other contexts.  Details.js is the monster now (1500 lines), and we could further split it out into mdraid/block/lvm2 plus support modules.

**Comment by mvollmer** [comment]
Squashed the FIXUPs.

**Comment by mvollmer** [comment]
> Let's wait for storaged 2.1.2, to hopefully fix the check-storage-mdraid failure.

Storaged 2.1.2 is in the images.

**Comment by mvollmer** [comment]
> Let's remove the HACK re storaged-project/storaged#12

Done.

**Comment by mvollmer** [comment]
> Why are the dialogs written in JSON/javascript?

We decided to go ahead with the current approach.

To clarify, there is no HTML or CSS in the JavaScript code.
Instead, there is one big Mustache template that can produce any
of the 19 storage dialogs, and the JavaScript controls the UI
strings, value validation, and visibility of fields.

This is quite clean because the dialogs are so similar and
primitive.  Once we polish them and they need to diverge, we can
make several templates, say.

**Comment by mvollmer** [comment]
Summary of open issues:

- There still is support for UDisks2 jobs in the code, but we
  might decide that we don't want that. Opinions?

  The idea is that on a system that also has UDisks2 and, say,
  GNOME Disks, we show the jobs that are running inside UDisks2,
  because it is nice.

- Curtains/blank-slate for the "No storage for you" page.

  I'll fix that.

- General time offset to peer.

  The concrete scenario is that Storaged jobs tell us their start time and estimated end time (relative to the clock on the server), and we want to show the estimated remaining time.  The easiest and most robust thing is to get a sample of the server clock and store the delta to the browser clock.  Then we can translate server time to equivalent browser time without making assumptions of all clocks being properly synchronized.

  This might be useful in all kinds of situations.

  I'd say we keep in mind to offer this from base1/cockpit somehow, but go with the current approach here.

**Comment by mvollmer** [comment]
> Curtains/blank-slate for the "No storage for you" page.

Fixed.

**Comment by mvollmer** [comment]
kubernetes is broken in Fedora 22 as of now and we picked it up when getting storaged 2.1.2.

**Comment by andreasn** [comment]
These looks a bit weird without, but I understand it's a good space for the spinner. Could we put the spinner on the right possibly? That would avoid it looking empty.

screenshot from 2015-07-16 15-36-29

**Comment by mvollmer** [comment]
> Could we put the spinner on the right possibly? That would avoid it looking empty.

Sure, this is your domain. :-)

I was thinking that we might want icons for arrays and volume groups as well.

**Comment by andreasn** [comment]
I got the format warning, but it still allowed me to do it, but then didn't succeed with the formatting in the end.
https://andreasn.portknox.net/index.php/s/BykI2y2CntSDTkx

**Comment by andreasn** [comment]
> I was thinking that we might want icons for arrays and volume groups as well.

There aren't any super established icons for these things, more than "blocks on top of each other":
screenshot
screenshot

I'm skeptical to using these icons as pure decoration, but it might be nice if they are able to provide additional information (like maybe degenerated RAID? But maybe that would be better as a pure warning icon).

(the regular hard drives below obviously doesn't convey any additional information right now either, so those are pure decoration right now too)

**Comment by mvollmer** [comment]
> I got the format warning, but it still allowed me to do it, but then didn't succeed with the formatting in the end.

Yes, good catch.  The formatting has failed, but we didn't show the error message because it happened in the background as a (very fast) "job", and we don't show job failures.  (That would be part of notifications, I'd say.)

I think the best immediate fix for this is to move the initial wipefs out of the background part: https://github.com/storaged-project/storaged/pull/15.  Then you will see it's errors right in the dialog.

We can also prevent people from hitting "Format" when there are usage alerts. I am not sure whether we can get them 100% accurate, though.

**Comment by mvollmer** [comment]
> I'm skeptical to using these icons as pure decoration,

Could you tweak the HTML/CSS until it looks acceptable?  I don't think you have to touch the JavaScript at all.

**Comment by mvollmer** [comment]
Ok, we have a workaround for the kube config boo-boo.

**Comment by andreasn** [comment]
So I've been pondering this some more, and I think the icon might be useful in the end after all, partly for identification and easier of scanning, and especially if we can have a unformatted-and-unpartioned-state of them (that I'll open a separate issue about).
Here is the icon: https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/graphics/storage-array.png

**Comment by mvollmer** [comment]
> Here is the icon: https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/graphics/storage-array.png

Thanks!  Should we use this also for volume groups?

**Comment by andreasn** [comment]
> Thanks! Should we use this also for volume groups?

Yes, it's all a collection of disks.

Mockup:
screenshot from 2015-07-16 18-28-44

**Comment by mvollmer** [comment]
> Mockup:

Implemented.  The spinner is now on the right.

**Pull request closed without merging**